### PR TITLE
Refactoring In-Progress

### PR DESCRIPTION
**Pull request by Suraj9696**
initial Refactoring Completed..
Base Component Removed..
Formatted all places.
Added camel case character
Removed all Underscore places 
Use Case/Mapper/Viewmodel - Separated
All AppConstant — Moved to gradle file - BuildConfig
Removed the comment in MarvelApplication 
Added missing dimensions in layouts
Updated the theme to Theme.MarvelApp
Added UTC For some classes
Updated - README.md file.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -103,6 +108,14 @@ dependencies {
     kapt 'com.google.dagger:dagger-compiler:2.31.2'
     kapt 'com.google.dagger:dagger-android-processor:2.31.2'
     implementation 'com.google.dagger:dagger-android-support:2.31.2'
+
+    testImplementation 'androidx.test:core:1.3.0'
+    // Optional -- Mockito framework
```
Comment (sarang-morde) [inline app/build.gradle:116]: remove all such comments

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -1,19 +1,21 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
     package="com.example.openbankassignment">
 
     <uses-permission android:name="android.permission.INTERNET" />
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
 
     <application
+        android:name=".MyApplication"
```
Comment (sarang-morde) [inline app/src/main/AndroidManifest.xml:10]: Please refactor this to MarvelApplication to keep it consistent

File: app/src/main/java/com/example/openbankassignment/base/BaseViewModel.kt [hunk 3]
```diff
@@ -2,23 +2,12 @@ package com.example.openbankassignment.base
 
 import androidx.lifecycle.ViewModel
 import io.reactivex.disposables.CompositeDisposable
-import io.reactivex.disposables.Disposable
 
 abstract class BaseViewModel : ViewModel() {
 
-    // TODO - Needs to add implementation
-
     private val compositeDisposable: CompositeDisposable = CompositeDisposable()
```
Comment (sarang-morde) [inline app/src/main/java/com/example/openbankassignment/base/BaseViewModel.kt:8]: Looks like there is no use for this variable. Please remove it if it's not required

File: app/src/main/java/com/example/openbankassignment/ui/adapter/HomeFragmentAdapter.kt [hunk 4]
```diff
@@ -11,48 +11,48 @@ import com.example.data.uimodel.CharacterModel
...
         }
     }
 
-    fun upDateDataSet(newList : List<CharacterModel>) {
+    fun upDateDataSet(newList: List<CharacterModel>) {
         characterModel = newList
         notifyDataSetChanged()
     }
 
+    interface HomeFragmentAdapterListener {
```
Comment (sarang-morde) [inline app/src/main/java/com/example/openbankassignment/ui/adapter/HomeFragmentAdapter.kt:49]: Please replace this interface with a lambda - 
e.g. private val mListener: ((CharacterModel) -> Unit)?

File: app/src/main/java/com/example/openbankassignment/MyApplication.kt [hunk 5]
```diff
@@ -8,7 +8,6 @@ class MyApplication : DaggerApplication() {
```
Comment (Akshata1812) [inline app/src/main/java/com/example/openbankassignment/MyApplication.kt:8]: Please change MyApplication to MarvelApplication

File: app/src/main/java/com/example/openbankassignment/viewmodel/CharacterDetailsViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+class CharacterDetailsViewModel @Inject constructor(
+    private var useCase: CharacterDetailsUseCase,
+    private var schedulers: SchedulerProvider
+) : BaseViewModel() {
+
+    private val compositeDisposable = CompositeDisposable()
+
+    private val characterDetail = MutableLiveData<CharacterModel>()
+    val characterDetails: LiveData<CharacterModel>
+        get() = characterDetail
```
Comment (sarang-morde) [inline app/src/main/java/com/example/openbankassignment/viewmodel/CharacterDetailsViewModel.kt:22]: This get() is not required. Please remove it.

File: app/src/main/java/com/example/openbankassignment/viewmodel/CharacterDetailsViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    override fun onCleared() {
+        compositeDisposable.clear()
+    }
+
+    fun getMarvelCharacter(characterId: Int) {
+        useCase.loadMarvelCharacterPage(characterId,
+            BuildConfig.PUBLIC_KEY, BuildConfig.PRIVATE_KEY,
+            System.currentTimeMillis()/1000)
```
Comment (sarang-morde) [inline app/src/main/java/com/example/openbankassignment/viewmodel/CharacterDetailsViewModel.kt:31]: Please move these params to the remoteSource

File: app/src/main/java/com/example/openbankassignment/viewmodel/CharacterViewModel.kt [hunk 8]
```diff
@@ -16,52 +15,32 @@ class CharacterViewModel @Inject constructor(
...
 
+    override fun onCleared() {
+        compositeDisposable.clear()
+    }
 
     fun getMarvelListData() {
-        useCase.loadMarvelCharacter(AppConstants.PUBLIC_KEY, AppConstants.PRIVATE_KEY,
-            System.currentTimeMillis()/1000)
+        useCase.loadMarvelCharacter(BuildConfig.PUBLIC_KEY, BuildConfig.PUBLIC_KEY,
+            System.currentTimeMillis() / 1000
```
Comment (sarang-morde) [inline app/src/main/java/com/example/openbankassignment/viewmodel/CharacterViewModel.kt:31]: Please move these params to the remoteSource

File: data/src/test/java/com/example/data/repo/MarvelRepositoryImplTest.kt [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Test
+    fun testGetMarvelCharacter() {
+        marvelRepository.getMarvelCharacter(CHARACTER_ID, PUBLIC_KEY, HASH, TIMESTAMP)
+        verify(characterApi).getCharactersDetails(CHARACTER_ID, PUBLIC_KEY, HASH, TIMESTAMP)
+    }
+
+    companion object {
+        const val PUBLIC_KEY = "public key"
```
Comment (Akshata1812) [inline data/src/test/java/com/example/data/repo/MarvelRepositoryImplTest.kt:40]: Please make then private

File: data/build.gradle [hunk 10]
```diff
@@ -52,8 +58,8 @@ dependencies {
     implementation "com.squareup.retrofit2:retrofit:2.9.0"
     implementation "com.squareup.retrofit2:converter-gson:2.9.0"
 
-   /* implementation "com.squareup.okhttp3:okhttp:4.9.0."
-    implementation "com.squareup.okhttp3:logging-interceptor:4.9.0"*/
+    /* implementation "com.squareup.okhttp3:okhttp:4.9.0."
```
Comment (sarang-morde) [inline data/build.gradle:61]: Please remove all such commented code

File: domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt [hunk 11]
```diff
@@ -0,0 +1,25 @@
+package com.example.domain.usecase.details
+
+import com.example.data.mapper.details.CharacterDetailsMapperImpl
+import com.example.data.repo.repository.MarvelRepository
+import com.example.data.uimodel.CharacterModel
+import com.example.domain.HashKeyGenerator
+import io.reactivex.Single
+import javax.inject.Inject
+
+class CharacterDetailsUseCaseImpl @Inject constructor(
```
Comment (Akshata1812) [inline domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt:10]: I think usecase test cases are missing

File: data/src/main/java/com/example/data/mapper/character/CharacterMapperImpl.kt [hunk 12]
```diff
@@ -0,0 +1,25 @@
+package com.example.data.mapper.character
+
+import com.example.data.entity.CharactersResponse
+import com.example.data.uimodel.CharacterModel
+import io.reactivex.Single
+
+class CharacterMapperImpl : CharacterMapper {
```
Comment (sarang-morde) [inline data/src/main/java/com/example/data/mapper/character/CharacterMapperImpl.kt:7]: Please add Unit tests for this class

File: data/src/main/java/com/example/data/mapper/details/CharacterDetailsMapperImpl.kt [hunk 13]
```diff
@@ -0,0 +1,16 @@
+package com.example.data.mapper.details
+
+import com.example.data.entity.CharactersResponse
+import com.example.data.uimodel.CharacterModel
+
+class CharacterDetailsMapperImpl : CharacterDetailsMapper {
```
Comment (sarang-morde) [inline data/src/main/java/com/example/data/mapper/details/CharacterDetailsMapperImpl.kt:6]: Please add Unit tests for this class

File: domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt (hunk 11 above)
Comment (sarang-morde) [inline domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt:10]: Please add Unit tests for this class

File: domain/src/main/java/com/example/domain/usecase/character/CharacterUseCaseImpl.kt [hunk 14]
```diff
@@ -10,29 +10,17 @@ import javax.inject.Inject
 /**
  * Implementation for Marvel character Use case Display
  */
-class CharacterUseCaseImpl @Inject constructor(private val marvelRepository: MarvelRepository
+class CharacterUseCaseImpl @Inject constructor(
```
Comment (sarang-morde) [inline domain/src/main/java/com/example/domain/usecase/character/CharacterUseCaseImpl.kt:13]: Please add Unit tests for this class

**Review by sarang-morde — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/openbankassignment/MyApplication.kt (hunk 5 above)
Comment (Suraj9696) [reply app/src/main/java/com/example/openbankassignment/MyApplication.kt:8]: Done

File: data/src/test/java/com/example/data/repo/MarvelRepositoryImplTest.kt (hunk 9 above)
Comment (Suraj9696) [reply data/src/test/java/com/example/data/repo/MarvelRepositoryImplTest.kt]: Done

File: domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt (hunk 11 above)
Comment (Suraj9696) [reply domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt:12]: Done

File: app/build.gradle (hunk 1 above)
Comment (Suraj9696) [reply app/build.gradle]: Done

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (Suraj9696) [reply app/src/main/AndroidManifest.xml]: Done

File: app/src/main/java/com/example/openbankassignment/base/BaseViewModel.kt (hunk 3 above)
Comment (Suraj9696) [reply app/src/main/java/com/example/openbankassignment/base/BaseViewModel.kt:8]: Done

File: app/src/main/java/com/example/openbankassignment/viewmodel/CharacterDetailsViewModel.kt (hunk 6 above)
Comment (Suraj9696) [reply app/src/main/java/com/example/openbankassignment/viewmodel/CharacterDetailsViewModel.kt]: Done

File: data/build.gradle (hunk 10 above)
Comment (Suraj9696) [reply data/build.gradle]: Done

File: data/src/main/java/com/example/data/mapper/character/CharacterMapperImpl.kt (hunk 12 above)
Comment (Suraj9696) [reply data/src/main/java/com/example/data/mapper/character/CharacterMapperImpl.kt:7]: Done

File: domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt (hunk 11 above)
Comment (Suraj9696) [reply domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt:12]: Done

File: domain/src/main/java/com/example/domain/usecase/character/CharacterUseCaseImpl.kt (hunk 14 above)
Comment (Suraj9696) [reply domain/src/main/java/com/example/domain/usecase/character/CharacterUseCaseImpl.kt:15]: Done

File: app/src/main/java/com/example/openbankassignment/base/BaseViewModel.kt (hunk 3 above)
Comment (sarang-morde) [reply app/src/main/java/com/example/openbankassignment/base/BaseViewModel.kt:8]: Is this `compositeDisposable` required? its a private variable and not being used anywhere in this file. Please remove it if it's not being used and then this file can be deleted as well.

File: app/src/main/java/com/example/openbankassignment/MarvelApplication.kt [hunk 15]
```diff
@@ -4,11 +4,10 @@ import com.example.openbankassignment.di.DaggerApplicationComponent
 import dagger.android.AndroidInjector
 import dagger.android.support.DaggerApplication
 
-class MyApplication : DaggerApplication() {
+class MarvelApplication : DaggerApplication() {
 
     override fun onCreate() {
```
Comment (sarang-morde) [inline app/src/main/java/com/example/openbankassignment/MarvelApplication.kt:9]: Please remove this empty onCreate fun

File: app/src/main/java/com/example/openbankassignment/ui/adapter/HomeFragmentAdapter.kt [hunk 16]
```diff
@@ -11,48 +11,45 @@ import com.example.data.uimodel.CharacterModel
...
 
+        holder.textView.text = currentUser.name
         holder.item.setOnClickListener {
-            mListener.onCharacterItemSelected(currentUser)
+            mListener.invoke(currentUser)
         }
     }
 
-    fun upDateDataSet(newList : List<CharacterModel>) {
+    fun upDateDataSet(newList: List<CharacterModel>) {
```
Comment (sarang-morde) [inline app/src/main/java/com/example/openbankassignment/ui/adapter/HomeFragmentAdapter.kt:44]: Please rename this as `updateDataSet
`

File: app/src/main/java/com/example/openbankassignment/ui/fragment/HomeFragment.kt [hunk 17]
```diff
@@ -4,64 +4,74 @@ import android.os.Bundle
...
 import androidx.navigation.fragment.findNavController
 import androidx.recyclerview.widget.LinearLayoutManager
 import com.example.data.uimodel.CharacterModel
 import com.example.openbankassignment.R
-import com.example.openbankassignment.base.BaseFragment
 import com.example.openbankassignment.ui.adapter.HomeFragmentAdapter
+import com.example.openbankassignment.viewmodel.CharacterDetailsViewModel
 import com.example.openbankassignment.viewmodel.CharacterViewModel
 import dagger.android.support.AndroidSupportInjection
 import kotlinx.android.synthetic.main.fragment_home.*
```
Comment (sarang-morde) [inline app/src/main/java/com/example/openbankassignment/ui/fragment/HomeFragment.kt:19]: Please avoid using Kotlin synthetics. Either use data binding or get the views using findviewbyid.

File: data/src/test/java/com/example/data/mapper/character/CharacterMapperImplTest.kt [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    @Before
+    fun setup() {
+        MockitoAnnotations.initMocks(this)
+        characterMapper = CharacterMapperImpl
+    }
+
+    @Test
+    fun testGetMarvelCharacterList() {
+        val characterResponse = CharactersResponse()
+        characterMapper.mapCharacterResponseToCharacter(characterResponse)
```
Comment (sarang-morde) [inline data/src/test/java/com/example/data/mapper/character/CharacterMapperImplTest.kt:24]: This UT is incomplete. Check that the response of `mapCharacterResponseToCharacter` fun is correct for given input.

File: data/src/test/java/com/example/data/mapper/details/CharacterDetailsMapperImplTest.kt [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    @Before
+    fun setup() {
+        MockitoAnnotations.initMocks(this)
+        characterDetailsMapper = CharacterDetailsMapperImpl
+    }
+
+    @Test
+    fun testGetMarvelCharacterList() {
+        val characterResponse = CharactersResponse()
+        characterDetailsMapper.mapCharacterDetailsResponse(characterResponse)
```
Comment (sarang-morde) [inline data/src/test/java/com/example/data/mapper/details/CharacterDetailsMapperImplTest.kt:24]: same here

**Review by Akshata1812 — APPROVED** [review]

File: data/src/main/java/com/example/data/mapper/details/CharacterDetailsMapperImpl.kt (hunk 13 above)
Comment (Suraj9696) [reply data/src/main/java/com/example/data/mapper/details/CharacterDetailsMapperImpl.kt]: Done

File: app/src/main/java/com/example/openbankassignment/ui/adapter/HomeFragmentAdapter.kt (hunk 4 above)
Comment (Suraj9696) [reply app/src/main/java/com/example/openbankassignment/ui/adapter/HomeFragmentAdapter.kt]: Done

File: app/src/main/java/com/example/openbankassignment/viewmodel/CharacterDetailsViewModel.kt (hunk 7 above)
Comment (Suraj9696) [reply app/src/main/java/com/example/openbankassignment/viewmodel/CharacterDetailsViewModel.kt]: Done

File: app/src/main/java/com/example/openbankassignment/viewmodel/CharacterViewModel.kt (hunk 8 above)
Comment (Suraj9696) [reply app/src/main/java/com/example/openbankassignment/viewmodel/CharacterViewModel.kt]: Done

File: domain/src/main/java/com/example/domain/usecase/character/CharacterUseCaseImpl.kt [hunk 20]
```diff
@@ -21,7 +21,7 @@ class CharacterUseCaseImpl @Inject constructor(
 
         return marvelRepository.getMarvelCharacterList(BuildConfig.PUBLIC_KEY, hash, timestamp())
             .map {
-                CharacterMapperImpl.mapCharacterResponseToCharacter(it)
+                CharacterMapperImpl().mapCharacterResponseToCharacter(it)
```
Comment (sarang-morde) [inline domain/src/main/java/com/example/domain/usecase/character/CharacterUseCaseImpl.kt:24]: Inject this mapper through the constructor

File: domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt [hunk 21]
```diff
@@ -22,7 +22,7 @@ class CharacterDetailsUseCaseImpl @Inject constructor(
             hash,
             timestamp()
         ).map {
-            CharacterDetailsMapperImpl.mapCharacterDetailsResponse(it)
+            CharacterDetailsMapperImpl().mapCharacterDetailsResponse(it)
```
Comment (sarang-morde) [inline domain/src/main/java/com/example/domain/usecase/details/CharacterDetailsUseCaseImpl.kt:25]: Same here

**Review by sarang-morde — APPROVED** [review]

**Pull request merged**